src/compositor/windowpixmapitem.cpp

```diff
@@ -480,13 +496,119 @@ QSGNode *WindowPixmapItem::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData
     SurfaceNode *node = static_cast<SurfaceNode *>(oldNode);
 
     if (m_item == 0) {
+        node->setTextureProvider(0, false);
         delete node;
         return 0;
     }
 
     if (!node) node = new SurfaceNode;
 
-    node->setTextureProvider(m_item->textureProvider());
+    class SnapshotTextureProvider : public QSGTextureProvider
+    {
+    public:
+        ~SnapshotTextureProvider()
+        {
+            delete fbo;
+            delete program;
+            delete t;
+        }
+        QSGTexture *texture() const Q_DECL_OVERRIDE
+        {
+            return t;
+        }
+        QSGTexture *t;
+        QOpenGLFramebufferObject *fbo;
+        QOpenGLShaderProgram *program;
+        int vertexLocation;
+        int sizeLocation;
+        int textureLocation;
+        int matrixLocation;
+    };
+
+    QSGTextureProvider *provider = m_item->textureProvider();
+    QSGTexture *texture = provider->texture();
+
+    if (!m_hasBuffer && texture) {
+        if (!m_textureProvider) {
+            SnapshotTextureProvider *prov = new SnapshotTextureProvider;
+            m_textureProvider = prov;
+
+            prov->fbo = 0;
+            prov->program = new QOpenGLShaderProgram;
+            prov->program->addShaderFromSourceCode(QOpenGLShader::Vertex,
+                "attribute highp vec4 vertex;\n"
+                "uniform highp mat4 matrix;\n"
+                "uniform highp vec2 size;\n"
+                "varying highp vec2 texPos;\n"
+                "void main(void) {\n"
+                "   texPos = vertex.xy;\n"
+                "   gl_Position = matrix * vec4(vertex.xy * size, vertex.zw);\n"
+                "}");
+            prov->program->addShaderFromSourceCode(QOpenGLShader::Fragment,
+                "uniform sampler2D texture;\n"
+                "varying highp vec2 texPos;\n"
+                "void main(void) {\n"
+                "   gl_FragColor = texture2D(texture, texPos);\n"
+                "}");
+            if (!prov->program->link())
+                qDebug() << prov->program->log();
+
+            prov->vertexLocation = prov->program->attributeLocation("vertex");
+            prov->sizeLocation = prov->program->uniformLocation("size");
+            prov->textureLocation = prov->program->uniformLocation("texture");
+            prov->matrixLocation = prov->program->uniformLocation("matrix");
+        }
+        provider = m_textureProvider;
+
+        if (m_unmapLock) {
+            SnapshotTextureProvider *prov = static_cast<SnapshotTextureProvider *>(provider);
+
+            if (!prov->fbo || prov->fbo->size() != QSize(width(), height())) {
+                delete prov->fbo;
+                prov->fbo = new QOpenGLFramebufferObject(width(), height());
+            }
+
+            prov->fbo->bind();
+            prov->program->bind();
+
+            glActiveTexture(GL_TEXTURE0);
```

Comment:
Probably, but maybe it's better to keep it explicit in case another item changes it in its updatePaintNode() if it is called before this one?